apps/builder/app/shared/nano-states/nano-states.ts

```diff
@@ -44,13 +47,84 @@ export const useSetTreeId = (treeId: Tree["id"]) => {
   });
 };
 
-export const rootInstanceContainer = atom<Instance | undefined>();
-export const useRootInstance = () => useValue(rootInstanceContainer);
-export const useSetRootInstance = (root: Instance) => {
+export const instancesStore = atom<Map<InstancesItem["id"], InstancesItem>>(
+  new Map()
+);
+export const useSetInstances = (
+  instances: [InstancesItem["id"], InstancesItem][]
+) => {
   useSyncInitializeOnce(() => {
-    rootInstanceContainer.set(root);
+    instancesStore.set(new Map(instances));
   });
 };
+
+export const patchInstancesMutable = (
+  rootInstance: undefined | Instance,
+  instances: Instances
+) => {
+  const oldInstancesIndex = createInstancesIndex(rootInstance);
+  const deletedInstanceIds = new Set<Instance["id"]>(instances.keys());
+  for (const oldInstance of oldInstancesIndex.instancesById.values()) {
+    deletedInstanceIds.delete(oldInstance.id);
+    const instance = instances.get(oldInstance.id);
+    const convertedOldInstance: InstancesItem = {
+      type: "instance",
+      id: oldInstance.id,
+      component: oldInstance.component,
+      children: oldInstance.children.map((child) => {
+        if (child.type === "text") {
+          return child;
+        }
+        return {
+          type: "id",
+          value: child.id,
+        };
+      }),
+    };
+    if (deepEqual(convertedOldInstance, instance)) {
```

Comment:
I am quite surprised to find a deepEqual, because the whole point of patches is to never have to do stuff like this.

Why did we not need this before?

Reply:
Because this is temporary solution to convert patches of denormalized tree to normalized patches. Just to not rewrite a lot of code. Will get rid of it step by step.

Reply:
got it, comment needed